In [1]:
import warnings

import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

import missingno as msno

import statsmodels.api as sm
import statsmodels.formula.api as smf

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.metrics import roc_curve, roc_auc_score

import factor_analyzer as fact

import funciones_auxiliares_hito_3 as aux3
from funciones_auxiliares_hito_2 import preprocesar_y_recodificar_enunciado_dos

# warnings.filterwarnings('ignore')
sns.set_theme()
plt.rcParams['figure.figsize'] = (12, 6)

In [2]:
df = pd.read_csv('students.csv', delimiter='|').drop(columns='Unnamed: 0')
df = preprocesar_y_recodificar_enunciado_dos(df)

# Analisis de factores latentes en variables famrel:health

In [3]:
posibles_factores_latentes = df.loc[:, 'famrel':'health']
print(fact.calculate_bartlett_sphericity(posibles_factores_latentes))

(264.4978629665067, 6.03816710126852e-48)


Por lo tanto, existe al menos una correlacion entre dos variables que es distinto de 0

In [4]:
fact.calculate_kmo(posibles_factores_latentes)

(array([0.50173343, 0.57692329, 0.61683967, 0.58664172, 0.5554092 ,
        0.59265861]),
 0.5767240731164592)

Como el valor del test de Kaiser Meyer Olkins es menor a 0.7, entonces NO existe un factor latente
en nuestros datos.

# Modelo para G1

In [7]:
modelo_g1 = aux3.crear_modelo_lineal(df.drop(columns=['G2', 'G3']), 'G1')
modelo_g1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     G1   R-squared:                       0.331
Model:                            OLS   Adj. R-squared:                  0.224
Method:                 Least Squares   F-statistic:                     3.090
Date:                Sat, 04 Feb 2023   Prob (F-statistic):           5.75e-08
Time:                        23:08:48   Log-Likelihood:                -683.63
No. Observations:                 284   AIC:                             1447.
Df Residuals:                     244   BIC:                             1593.
Df Model:                          39                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept          9.6531      3.452      2.797      0.006       2.855      16.452
school            -0.3886      0.666     -0.583      0.560      -1.701       0.924
sex                0.7631      0.432      1.768      0.078      -0.087       1.613
age                0.0772      0.185      0.418      0.677      -0.287       0.441
address           -0.0205      0.487     -0.042      0.966      -0.980       0.939
famsize            0.3125      0.440      0.710      0.478      -0.555       1.180
Pstatus            0.1872      0.650      0.288      0.774      -1.094       1.468
Medu               0.1809      0.282      0.642      0.521      -0.374       0.736
Fedu               0.1527      0.236      0.646      0.519      -0.313       0.618
traveltime        -0.0215      0.291     -0.074      0.941      -0.594       0.551
studytime          0.5074      0.245      2.068      0.040       0.024       0.991
failures          -1.1235      0.282     -3.977      0.000      -1.680      -0.567
schoolsup         -1.7737      0.564     -3.145      0.002      -2.885      -0.663
famsup             1.0062      0.411      2.447      0.015       0.196       1.816
paid               0.0321      0.403      0.080      0.937      -0.762       0.826
activities         0.4719      0.370      1.276      0.203      -0.257       1.201
nursery            0.0255      0.455      0.056      0.955      -0.871       0.922
higher            -1.1077      0.876     -1.265      0.207      -2.833       0.617
internet          -0.0671      0.502     -0.134      0.894      -1.055       0.921
romantic          -0.0842      0.399     -0.211      0.833      -0.870       0.702
famrel             0.1449      0.213      0.681      0.497      -0.274       0.564
freetime           0.3778      0.193      1.959      0.051      -0.002       0.758
goout             -0.7691      0.199     -3.873      0.000      -1.160      -0.378
Dalc              -0.0730      0.286     -0.255      0.799      -0.636       0.490
Walc               0.1585      0.210      0.756      0.451      -0.255       0.572
health            -0.2918      0.142     -2.062      0.040      -0.571      -0.013
absences          -0.0048      0.024     -0.196      0.845      -0.053       0.043
health_Mjob        1.0251      0.918      1.116      0.265      -0.784       2.834
other             -0.5423      0.598     -0.907      0.365      -1.720       0.636
services           0.3913      0.676      0.579      0.563      -0.939       1.722
teacher           -1.0850      0.836     -1.298      0.195      -2.731       0.561
health_Fjob       -1.0572      1.304     -0.810      0.418      -3.626       1.512
other_Fjob        -1.4765      0.984     -1.500      0.135      -3.415       0.462
services_Fjob     -0.8524      0.990     -0.861      0.390      -2.801       1.097
teacher_Fjob

Al realizar un modelo de regresion lineal con todas las variables y G1 como vector objetivo se
obtuvo lo siguiente:

- El modelo presenta un $R^2$ = 0.331. Por lo tanto, las variables explican un 33.1% de la variacion
de ```G1```
- Solo el intercepto, ```studytime```, ```failures```, ```schoolsup```, ```famsup```, ```goout``` y ```health``` afectan significativamente a la variable ```G1``` 